packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -215,45 +249,94 @@ async function visualizeObjectNode(
+): Promise<VisualSharedTreeNode> {
+	const mapNodeSchemaProperties: Record<string, FieldSchemaProperties> = {};
+	for (const key of Object.keys(tree.fields)) {
+		mapNodeSchemaProperties[key] = {
+			allowedTypes: schema.allowedTypes,
+			// Map values are always required. Don't display field requirement information, since that information is redundant.
+			isRequired: undefined,
+		};
+	}
+
+	return {
+		schema: {
+			schemaName: tree.type,
+			allowedTypes: concatenateTypes(allowedTypes ?? new Set()),
+			isRequired: isRequired?.toString(),
+		},
+		fields: await visualizeVerboseNodeFields(
+			tree.fields,
+			treeDefinitions,
+			mapNodeSchemaProperties,
+			visualizeChildData,
+		),
+		kind: VisualSharedTreeNodeKind.InternalNode,
+	};
+}
+
+/**
+ * Returns the schema & fields of the node with type {@link ArrayNodeStoredSchema}.
+ */
+async function visualizeArrayNode(
+	tree: VerboseTreeNode,
+	schema: SimpleArrayNodeSchema,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	{ allowedTypes, isRequired }: FieldSchemaProperties,
 	visualizeChildData: VisualizeChildData,
 ): Promise<VisualSharedTreeNode> {
+	const children = tree.fields;
+	if (!Array.isArray(children)) {
+		throw new TypeError("Invalid array");
+	}
+
+	const arrayNodeSchemaProperties: Record<string, FieldSchemaProperties> = {};
+	for (const [i] of children.entries()) {
+		arrayNodeSchemaProperties[i] = {
+			allowedTypes: schema.allowedTypes,
+			isRequired: undefined,
+		};
+	}
+
 	return {
 		schema: {
 			schemaName: tree.type,
-			allowedTypes: `Record<string, ${concatenateTypes(nodeSchema.allowedTypes)}>`,
+			allowedTypes: concatenateTypes(allowedTypes ?? new Set()),
+			isRequired: isRequired?.toString(),
 		},
-		fields: await visualizeVerboseNodeFields(tree, treeSchema, visualizeChildData),
+		fields: await visualizeVerboseNodeFields(
+			tree.fields,
+			treeDefinitions,
+			arrayNodeSchemaProperties,
+			visualizeChildData,
+		),
 		kind: VisualSharedTreeNodeKind.InternalNode,
 	};
 }
 
 /**
- * Main recursive helper function to create the visual representation of the SharedTree.
- * Processes tree nodes based on their schema type (e.g., ObjectNodeStoredSchema, MapNodeStoredSchema, LeafNodeStoredSchema), producing the visual representation for each type.
+ * Creates the visual representation of non-leaf SharedTree nodes.
+ *
+ * @remarks
+ * Processes internal tree nodes based on their schema type (e.g., ObjectNodeStoredSchema, MapNodeStoredSchema, ArrayNodeStoredSchema),
+ * producing the visual representation for each type.
  *
  * @see {@link https://fluidframework.com/docs/data-structures/tree/} for more information on the SharedTree schema.
  *
  * @remarks
  */
-export async function visualizeSharedTreeNodeBySchema(
-	tree: VerboseTree,
-	treeSchema: SimpleTreeSchema,
+async function visualizeNodeBySchema(
+	tree: VerboseTreeNode,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	{ allowedTypes, isRequired }: FieldSchemaProperties,
```

Comment:
I don't think its a good pattern to go passing in these field schema properties to each node visualizer: that makes you have to handle them in many different places.

Instead I'd recommend writing a single function that handles field related things, maybe call it `visualizeField` and have it take in the field ( `VerboseTree | undefined`) and its schema ( `SimpleFieldSchema` ).

Use that to handle the root, and every other field, and you should be able to remove a lot of unnecessary duplication from the visualizer code.